app/reducers/rootReducer.ts

```diff
@@ -1,94 +1,140 @@
 import { POPUP_PROPS, Action } from "../types";
+import * as actionType from "../actions";
+
 export const INITIAL_POPUP_STATE: POPUP_PROPS = {
-  enabled: false,
-  tabId: -1,
-  tabUrl: "",
-  errorMessage: "",
+  data: {},
+  currentUrl: "",
+  currentTab: -1
+};
+
+const initialTabProperties = {
+  enabledStatus: false,
   requests: [],
+  interceptStatus: "",
+  errorMessage: "",
+  PageDetails: [],
   checkedReqs: {},
-  responseText: {},
-  statusCodes: {},
   contentType: {},
-  PageDetails: [],
-  interceptStatus: "",
-  isInterceptorOn: {},
+  isInterceptorOn: true,
   responseData: {},
-  responseError: {}
+  responseError: {},
+  responseText: {},
+  statusCodes: {}
 };
 
-//ACTION CONSTANTS
-import * as actionType from "../actions";
+function extendStateData(state: POPUP_PROPS, payload: any, props: any) {
+  return {
+    ...state,
+    data: {
+      ...state.data,
+      [payload.tabId]: {
```

Comment:
I think here we using only the `tabId` from payload.
But we can simplify this by avoiding the `props` and using the `payload` alone.

```
const { tabId, ...otherPayloadItems } = payload
return {
    ...state,
    data: {
      ...state.data,
      [tabId]: {
        ...state.data[tabId],
        ...otherPayloadItems,
     }
  }
}
```

This way we can avoid the whole `props` argument.

Reply:
This can be done. But this only does a shallow merging of objects and not deep merging. This works for actions such as `UPDATE_REQUESTS` and `TOGGLE_LISTENING`, but we need deep merging of objects in actions like `RESP_TEXT_CHANGE` and `TOGGLE_CHECKBOX`. Any other way we could achieve that?

Reply:
Only for those two do the deep merge in reducer and send the new object as payload to this method.